spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/CompositeConfiguration.java

```diff
@@ -32,6 +32,7 @@
  * @author Ryan Baxter
  */
 @Configuration
+@ConditionalOnMissingBean(CompositeEnvironmentRepository.class)
```

Comment:
Actually I think the problem is below on the `compositeEncironmentRepository` method.  Looks like I added the wrong conditional class there.

Reply:
Both bean methods in this class produce `CompositeEnvironmentRepository` beans (it just happens one is the more specific type `SearchPathCompositeEnvironmentRepository`). If the condition were moved to the `compositeEnvironmentRepository()` method, then users would still be unable to provide a primary `CompositeEnvironmentRepository` if the `SearchPathCompositeEnvironmentRepository` bean is made.

If you change the test configuration `OverrideCompositeConfig` to the following, you can see that happen (because the `MultipleJGitEnvironmentRepository` bean is a `SearchPathLocator` bean):

```java
@Bean
@Primary
CompositeEnvironmentRepository customCompositeEnvironmentRepository() {
	return new CompositeEnvironmentRepository(Arrays.<EnvironmentRepository>asList(new TestOrderedEnvironmentRepository(1, new Environment("app", "dev"), null), someGitRepo()));
}

@Bean
MultipleJGitEnvironmentRepository someGitRepo() {
	MultipleJGitEnvironmentRepository gitRepo = new MultipleJGitEnvironmentRepository(new MockEnvironment());
	gitRepo.setUri("some-uri");
	return gitRepo;
}
```

Without guarding beans marked primary with some `ConditionalOnMissingBean`, the only option for users wanting a different primary bean is excluding the configuration that creates them, as far as I know. In this case, that means excluding the `ConfigServerAutoConfiguration`.

Reply:
Ah yes, I forgot they were subclasses of each other makes sense.